### ISSUE TITLE
Model surgery on linear layer's features

### ISSUE THREAD

**Issue by ddrous** [patrick-kidger/equinox#459]
Hi @patrick-kidger,

So I'm performing model surgery, and so far I've been able to infer the `in_features` of a linear layer by the weight's shape. But now I have to modify, in addition to the Jax learnable arrays, the integers for the in_ and out_ features, but Equinox is returning a not very informative error:

``` python
import jax
import equinox as eqx

key = jax.random.PRNGKey(42)

layer = eqx.nn.Linear(10, 100, key=key)
print(layer.in_features)

where = lambda l: l.weight
new_layer = eqx.tree_at(where, layer, layer.weight)
print(new_layer)    ## This works

where = lambda l: l.in_features
new_layer = eqx.tree_at(where, layer, 11)
print(new_layer)    ## This doesn't work !
```

Any suggestion on how to get around this issue ? Thanks in advance : )

**P.S.** sorry I'm little lazy to dig through the code; and also I'm sure I'll learn more than I can imagine by learning directly from the makers.

**Comment by patrick-kidger** [comment]
So the answer is that this isn't really defined/supported. `layer.in_features` isn't part of the pytree structure -- it's just treated as extra arbitrary metadata. (It's morally equivalent to `dict().get` -- you don't expect to be able to overwrite the method on a dictionary, either.) Try doing `jax.tree_util.tree_leaves(layer)`: the weight and bias are printed out, but not any of the extra metadata, which aren't leaves.

Can you create a brand-new linear layer with the appropriate sizes, and then do `eqx.tree_at` on that?

**Comment by ddrous** [comment]
Ohh interesting. Thanks a lot. 

Well, I tried creating a brand new linear layer with appropriate sizes, it worked for MLPs as suggested in the [docs](https://docs.kidger.site/equinox/tricks/#performing-model-surgery). But I ran into trouble when the architecture of the network was kinda arbitrary with no, say, `layers` attribute.

Also, this isn't really important, but would it improve things if such features were wrapped into a `jax.Array`, made part of the PyTree structure, and no gradients propagated back to them ?
